drivers/mdio/mdio_nxp_enet.c

```diff
@@ -0,0 +1,297 @@
+static int nxp_enet_mdio_write(const struct device *dev,
+			uint8_t prtad, uint8_t regad, uint16_t write_data)
+{
+	const struct nxp_enet_mdio_config *config = dev->config;
+	struct nxp_enet_mdio_data *data = dev->data;
+	int ret;
+
+	/* Only one MDIO bus operation attempt at a time */
+	(void)k_mutex_lock(&data->mdio_mutex, K_FOREVER);
+
+	/*
+	 * Clear the bit (W1C) that indicates MDIO transfer is ready to
+	 * prepare to wait for it to be set once this write is done
+	 */
+	config->base->EIR |= ENET_EIR_MII_MASK;
+
+	/*
+	 * Write MDIO frame to MII management register which will
+	 * send the write command and data out to the MDIO bus as this frame:
+	 * ST = start, 1 means start
+	 * OP = operation, 1 means write
+	 * PA = PHY/Port address
+	 * RA = Register/Device Address
+	 * TA = Turnaround, must be 2 to be valid
+	 * data = data to be written to the PHY register
+	 */
+	config->base->MMFR = ENET_MMFR_ST(0x1U) |
+				ENET_MMFR_OP(MDIO_OP_C22_WRITE) |
+				ENET_MMFR_PA(prtad) |
+				ENET_MMFR_RA(regad) |
+				ENET_MMFR_TA(0x2U) |
+				write_data;
+
+	ret = nxp_enet_mdio_wait_xfer(dev);
+	if (ret) {
+		(void)k_mutex_unlock(&data->mdio_mutex);
+		return ret;
+	}
+
+	/* Clear the same bit as before because the event has been handled */
+	config->base->EIR |= ENET_EIR_MII_MASK;
+
+	/* This MDIO interaction is finished */
+	(void)k_mutex_unlock(&data->mdio_mutex);
+
+	return ret;
+}
+
+/* MDIO bus enable/disable "implementation" */
+static void nxp_enet_mdio_bus_fn(const struct device *dev)
+{
+	/*
+	 * MDIO bus device is actually part of ethernet device, and
+	 * does not support ability to disable/enable MDIO bus hardware
+	 * independently of the ethernet/MAC hardware, so do nothing.
+	 */
+}
+
+static const struct mdio_driver_api nxp_enet_mdio_api = {
+	.read = nxp_enet_mdio_read,
+	.write = nxp_enet_mdio_write,
+	.bus_enable = nxp_enet_mdio_bus_fn,
+	.bus_disable = nxp_enet_mdio_bus_fn,
+};
+
+static void nxp_enet_mdio_isr_cb(const struct device *dev)
+{
+	const struct nxp_enet_mdio_config *config = dev->config;
+	struct nxp_enet_mdio_data *data = dev->data;
+
+	/* Signal that operation finished */
+	k_sem_give(&data->mdio_sem);
+
+	/* Disable the interrupt */
+	config->base->EIMR &= ~ENET_EIMR_MII_MASK;
+}
+
+static void nxp_enet_mdio_post_module_reset_init(const struct device *dev)
+{
+	const struct nxp_enet_mdio_config *config = dev->config;
+	uint32_t enet_module_clock_rate;
+
+	/* Set up MSCR register */
+	(void) clock_control_get_rate(config->clock_dev, config->clock_subsys,
+							&enet_module_clock_rate);
+	uint32_t mii_speed = (enet_module_clock_rate + 2 * NXP_ENET_MDIO_MDC_FREQ - 1) /
+					(2 * NXP_ENET_MDIO_MDC_FREQ) - 1;
+	uint32_t holdtime = (10 + NSEC_PER_SEC / enet_module_clock_rate - 1) /
+					(NSEC_PER_SEC / enet_module_clock_rate) - 1;
+	uint32_t mscr = ENET_MSCR_MII_SPEED(mii_speed) | ENET_MSCR_HOLDTIME(holdtime) |
+			(config->disable_preamble ? ENET_MSCR_DIS_PRE_MASK : 0);
+	config->base->MSCR = mscr;
```

Comment:
Is there no need for lock and unlock the mutex while writing to the MSCR register?

Reply:
no, the mutex is for bus operation, which is initiated by functions part of the public api, so needs a mutex. this function just is called once for the module and is a private driver api, and even if multiple threads somehow call it (which should not happen) it doesn't matter because this is the only value ever written to this register.